File: services/svc-api-meta/cmd/server/registry.go
// Package main: console sub-app registry (02-frontend-architecture.md §4.1).
//
// The registry is a backend-served, frontend-consumed dynamic list of which
// sub-apps the console shell should load. The shell fetches it at boot
// (02§4.2) and caches in sessionStorage (5 min), falling back to a built-in
// list on failure.
//
// Production source: Nacos config + MySQL mf_app_registry rows, managed by an
// internal ops console. Phase-1 keeps an in-memory store seeded with the
// canonical app set so the frontend pulls a real endpoint (no hardcoded
// fallback in the shell).
package main

import (
	"encoding/json"
	"net/http"
	"sync"
)

// RegistryApp mirrors the JSON the console shell expects (02§4.1 API example).
type RegistryApp struct {
	AppCode      string         `json:"appCode"`
	AppTitle     string         `json:"appTitle"`
	ProductCodes []string       `json:"productCodes"`
	ActiveRules  []string       `json:"activeRules"`
	EntryURL     string         `json:"entryUrl"`
	Version      string         `json:"version"`
	KeepAlive    bool           `json:"keepAlive"`
	Preload      bool           `json:"preload"`
	Status       string         `json:"status"`
	Gray         *RegistryGray  `json:"gray,omitempty"`
}

type RegistryGray struct {
	Version  string         `json:"version"`
	EntryURL string         `json:"entryUrl"`
	Rule     map[string]any `json:"rule"`
}

// Registry is the full response: { apps, revision, ttl }.
type Registry struct {
	Apps    []RegistryApp `json:"apps"`
	Revision int          `json:"revision"`
	TTL      int          `json:"ttl"`
}

// registryStore is the in-memory registry (MySQL + Nacos in production).
type registryStore struct {
	mu       sync.RWMutex
	registry Registry
}

func newRegistryStore() *registryStore {
	return &registryStore{registry: seedRegistry()}
}

// seedRegistry returns the canonical phase-1 sub-app set. EntryURLs point at
// the dev servers; production swaps these for versioned static-domain entries
// (https://static.starcloud.cn/apps/{app}/{ver}/index.html).
func seedRegistry() Registry {
	return Registry{
		Revision: 1,
		TTL:      300,
		Apps: []RegistryApp{
			{AppCode: "console-ecs", AppTitle: "云服务器 ECS", ProductCodes: []string{"scecs"}, ActiveRules: []string{"/scecs"}, EntryURL: "http://localhost:5174/", Version: "0.1.0", KeepAlive: true, Preload: true, Status: "online"},
			{AppCode: "console-storage", AppTitle: "对象存储 OSS", ProductCodes: []string{"scoss"}, ActiveRules: []string{"/scoss"}, EntryURL: "http://localhost:5176/", Version: "0.1.0", KeepAlive: true, Preload: false, Status: "online"},
			{AppCode: "console-network", AppTitle: "专有网络 VPC", ProductCodes: []string{"scvpc", "sceip"}, ActiveRules: []string{"/scvpc", "/sceip"}, EntryURL: "http://localhost:5177/", Version: "0.1.0", KeepAlive: false, Preload: false, Status: "online"},
			{AppCode: "console-database", AppTitle: "云数据库 RDS", ProductCodes: []string{"scrds"}, ActiveRules: []string{"/scrds"}, EntryURL: "http://localhost:5178/", Version: "0.1.0", KeepAlive: false, Preload: false, Status: "online"},
			{AppCode: "console-monitor", AppTitle: "云监控", ProductCodes: []string{"scmon"}, ActiveRules: []string{"/scmon"}, EntryURL: "http://localhost:5179/", Version: "0.1.0", KeepAlive: false, Preload: false, Status: "online"},
			{AppCode: "web-billing", AppTitle: "费用中心", ProductCodes: []string{}, ActiveRules: []string{"/billing"}, EntryURL: "http://localhost:5180/", Version: "0.1.0", KeepAlive: false, Preload: true, Status: "online"},
			{AppCode: "web-account", AppTitle: "账号与访问控制", ProductCodes: []string{}, ActiveRules: []string{"/account"}, EntryURL: "http://localhost:5175/", Version: "0.1.0", KeepAlive: false, Preload: false, Status: "online"},
			{AppCode: "web-ticket", AppTitle: "工单支持", ProductCodes: []string{}, ActiveRules: []string{"/ticket"}, EntryURL: "http://localhost:5181/", Version: "0.1.0", KeepAlive: false, Preload: false, Status: "online"},
			{AppCode: "devops-explorer", AppTitle: "OpenAPI Explorer", ProductCodes: []string{}, ActiveRules: []string{"/explorer"}, EntryURL: "http://localhost:5182/", Version: "0.1.0", KeepAlive: false, Preload: false, Status: "online"},
		},
	}
}

// handleConsoleApps returns the sub-app registry. This endpoint is public to
// the console shell (fetched in parallel with /api/iam/session at boot, before
// login may be established, 02§4.2 sequence) — no account header required.
// ETag + Cache-Control: no-cache so the shell can revalidate cheaply.
func (s *registryStore) handleConsoleApps(w http.ResponseWriter, r *http.Request) {
	s.mu.RLock()
	reg := s.registry
	s.mu.RUnlock()

	// The shell reads revision for change detection (02§4.1).
	body := map[string]any{
		"RequestId": w.Header().Get("X-Sc-TraceId"),
		"Code":      "OK",
		"Data":      reg,
	}
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-cache")
	_ = json.NewEncoder(w).Encode(body)
}

File: platform/frontend/apps/console-base/src/registry.ts
/**
 * Sub-app registry client (02§4.1).
 * Source of truth is svc-api-meta GET /api/v1/meta/console/apps; this falls
 * back to a built-in static list so the shell runs with zero backend.
 * Cached in sessionStorage for 5 minutes (02§4.1).
 */
import { defineStore } from "pinia";

export interface RegistryApp {
  appCode: string;
  appTitle: string;
  productCodes: string[];
  activeRules: string[];
  entryUrl: string;
  version: string;
  keepAlive: boolean;
  preload: boolean;
  gray?: { version: string; entryUrl: string; rule: { type: string; percent?: number } };
  status: string;
}

export interface Registry {
  apps: RegistryApp[];
  revision: number;
  ttl: number;
}

const CACHE_KEY = "sc:console:registry";
const CACHE_TTL = 5 * 60 * 1000;

// Built-in fallback (production pulls this from svc-api-meta). Mirrors the
// architecture doc's seven product-category sub-apps + billing/iam/ticket.
const fallback: Registry = {
  revision: 1,
  ttl: 300,
  apps: [
    { appCode: "console-ecs", appTitle: "云服务器 ECS", productCodes: ["scecs"], activeRules: ["/scecs"], entryUrl: "http://localhost:5174/", version: "0.1.0", keepAlive: true, preload: true, status: "online" },
    { appCode: "console-eci", appTitle: "弹性容器实例 ECI", productCodes: ["sceci"], activeRules: ["/sceci"], entryUrl: "http://localhost:5183/", version: "0.1.0", keepAlive: true, preload: false, status: "online" },
    { appCode: "console-lb", appTitle: "负载均衡 SLB", productCodes: ["sclb"], activeRules: ["/sclb"], entryUrl: "http://localhost:5184/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-autoscaling", appTitle: "弹性伸缩 AS", productCodes: ["scas"], activeRules: ["/scas"], entryUrl: "http://localhost:5185/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-backup", appTitle: "云备份", productCodes: ["scbackup"], activeRules: ["/scbackup"], entryUrl: "http://localhost:5186/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-redis", appTitle: "云数据库 Redis", productCodes: ["scredis"], activeRules: ["/scredis"], entryUrl: "http://localhost:5187/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-kafka", appTitle: "消息队列 Kafka", productCodes: ["sckafka"], activeRules: ["/sckafka"], entryUrl: "http://localhost:5188/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-logservice", appTitle: "日志服务", productCodes: ["sclog"], activeRules: ["/sclog"], entryUrl: "http://localhost:5189/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-storage", appTitle: "对象存储 OSS", productCodes: ["scoss"], activeRules: ["/scoss"], entryUrl: "http://localhost:5176/", version: "0.1.0", keepAlive: true, preload: false, status: "online" },
    { appCode: "console-network", appTitle: "专有网络 VPC", productCodes: ["scvpc", "sceip"], activeRules: ["/scvpc", "/sceip"], entryUrl: "http://localhost:5177/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-database", appTitle: "云数据库 RDS", productCodes: ["scrds"], activeRules: ["/scrds"], entryUrl: "http://localhost:5178/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "console-monitor", appTitle: "云监控", productCodes: ["scmon"], activeRules: ["/scmon"], entryUrl: "http://localhost:5179/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "web-billing", appTitle: "费用中心", productCodes: [], activeRules: ["/billing"], entryUrl: "http://localhost:5180/", version: "0.1.0", keepAlive: false, preload: true, status: "online" },
    { appCode: "web-account", appTitle: "账号与访问控制", productCodes: [], activeRules: ["/account"], entryUrl: "http://localhost:5175/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "web-ticket", appTitle: "工单支持", productCodes: [], activeRules: ["/ticket"], entryUrl: "http://localhost:5181/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
    { appCode: "devops-explorer", appTitle: "OpenAPI Explorer", productCodes: [], activeRules: ["/explorer"], entryUrl: "http://localhost:5182/", version: "0.1.0", keepAlive: false, preload: false, status: "online" },
  ],
};

export const useRegistry = defineStore("registry", {
  state: () => ({
    apps: [] as RegistryApp[],
    revision: 0,
    loaded: false,
  }),
  actions: {
    async load() {
      // 1) Session cache (5 min).
      try {
        const raw = sessionStorage.getItem(CACHE_KEY);
        if (raw) {
          const cached = JSON.parse(raw) as { data: Registry; ts: number };
          if (Date.now() - cached.ts < CACHE_TTL) {
            this.apply(cached.data);
            return;
          }
        }
      } catch { /* ignore corrupt cache */ }

      // 2) Runtime registry from svc-api-meta. The service returns the
      // platform envelope { Code, Data: { apps, revision, ttl } } (03§9.3);
      // unwrap Data before applying.
      try {
        const res = await fetch("/api/v1/meta/console/apps");
        if (res.ok) {
          const body = await res.json();
          const data = (body.Data ?? body) as Registry;
          if (data && Array.isArray(data.apps)) {
            this.apply(data);
            sessionStorage.setItem(CACHE_KEY, JSON.stringify({ data, ts: Date.now() }));
            return;
          }
        }
      } catch (e) {
        console.warn("[console-base] registry fetch failed, using fallback", e);
      }

      // 3) Built-in fallback (02§4.1: 保底可用).
      this.apply(fallback);
    },
    apply(data: Registry) {
      this.apps = Array.isArray(data?.apps) ? data.apps : [];
      this.revision = data?.revision ?? 0;
      this.loaded = true;
    },
    /** Resolve which sub-app owns a given path (first path segment match). */
    resolve(path: string): RegistryApp | undefined {
      const seg = "/" + (path.split("/").filter(Boolean)[0] ?? "");
      return this.apps.find((a) => a.activeRules.includes(seg));
    },
    productCodeOf(path: string): string | undefined {
      const seg = "/" + (path.split("/").filter(Boolean)[0] ?? "");
      return this.resolve(path)?.activeRules.find((r) => r === seg);
    },
  },
});
